Cache literal ranges constructed from immediate values

headius opened the pull request:
This adds logic in IR and JIT to cache the resulting Range object when a literal range is composed of other immediate literal values, like longs.

The first pass here is a simple introduction of logic to cache the resulting object, while still emitting code to create the input values. A more compact representation would not need to revisit the input value logic (even to check a cache) and would instead embed the values directly into the range, so it can be constructed and cached outside of IR/JIT output.

enebo commented:
SHIP IT

enebo commented on core/src/main/java/org/jruby/ir/operands/Range.java:
```diff
@@ -0,0 +1,82 @@
+
+    @Override
+    public OperandType getOperandType() {
+        return OperandType.RANGE;
```
I don't see OperandType here?

enebo replied:
Oh it is already there...what! :)

enebo replied:
Ah ok this was originally an Operand and we realized since it could raise we had to make it buildRange.  I wonder why we never had the discussion of both since we can check to see if it will raise during IRBuild in cases?

enebo reviewed (approved):
I am not sure how I entered review mode but I had a question and answered it so I have no issues with anything here.  Seems like a reasonable base to build on top of too.

kares commented on core/src/main/java/org/jruby/ir/targets/simple/NormalValueCompiler.java:
```diff
@@ -125,6 +127,16 @@ public void pushByteList(final ByteList bl) {
+            begin.run();
+            end.run();
+            compiler.adapter.pushBoolean(exclusive);
+            compiler.adapter.invokestatic(p(RubyRange.class), "newRegexp", sig(RubyRange.class, ThreadContext.class, IRubyObject.class, IRubyObject.class, boolean.class));
```
the RubyRange factory method name (`newRegexp`) seems like a copy pasta typo

headius replied:
And here I was thinking "huh, maybe we need a better way to test the non-indy logic for constant values". Will fix and add a way to test.

headius replied:
Adding an AOT mode to the compiler's `general_spec.rb` would have found this:

```
  10) JRuby's AOTSpecUtils assigns literal values to locals
      Failure/Error: Unable to infer file and line number from backtrace
      
      Java::JavaLang::NoSuchMethodError:
        'org.jruby.RubyRange org.jruby.RubyRange.newRegexp(org.jruby.runtime.ThreadContext, org.jruby.runtime.builtin.IRubyObject, org.jruby.runtime.builtin.IRubyObject, boolean)'
      # Users.headius.projects.jruby.spec.compiler.general_spec.range0(Users/headius/projects/jruby/spec/compiler//Users/headius/projects/jruby/spec/compiler/general_spec.rb)
      # Users.headius.projects.jruby.spec.compiler.general_spec.<main>(/Users/headius/projects/jruby/spec/compiler/general_spec.rb:170)
      # jdk.internal.reflect.GeneratedMethodAccessor22.invoke(jdk/internal/reflect/GeneratedMethodAccessor22)
      # jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(jdk/internal/reflect/DelegatingMethodAccessorImpl.java:43)
      # java.lang.reflect.Method.invoke(java/lang/reflect/Method.java:566)
      # RUBY.compile_run(/Users/headius/projects/jruby/spec/compiler/general_spec.rb:129)
      # RUBY.run(/Users/headius/projects/jruby/spec/compiler/general_spec.rb:74)
      # RUBY.<main>(/Users/headius/projects/jruby/spec/compiler/general_spec.rb:170)
```

headius commented:
Addressed all review comments and added a test run to ensure the non-indy side of this (and all other compilation logic) gets tested. Bonus: a couple fixes found by adding AOT to the compiler spec run.

headius commented:
Also note: a more generic solution would advise a way to construct large chains of constant literals, caching where appropriate without emitting IR or bytecode that only runs once. With the increasing interest in immutable objects, I expect we will see more complex ways to initialize immutable arrays, hashes, and structs.
